core/src/trezor/crypto/slip39.py

```diff
@@ -0,0 +1,606 @@
+
+    if len(group_thresholds) != 1:
+        raise MnemonicError(
+            "Invalid set of mnemonics. All mnemonics must have the same group threshold."
+        )
+
+    if len(group_counts) != 1:
+        raise MnemonicError(
+            "Invalid set of mnemonics. All mnemonics must have the same group count."
+        )
+
+    for group_index, group in groups.items():
+        if len(set(share[0] for share in group[1])) != len(group[1]):
+            raise MnemonicError(
+                "Invalid set of shares. Member indices in each group must be unique."
+            )
+
+    return (
+        identifiers.pop(),
+        iteration_exponents.pop(),
+        group_thresholds.pop(),
+        group_counts.pop(),
+        groups,
+    )
+
+
+def _generate_random_identifier():
+    """Returns a randomly generated integer in the range 0, ... , 2**ID_LENGTH_BITS - 1."""
+
+    identifier = int.from_bytes(random.bytes(bits_to_bytes(ID_LENGTH_BITS)), "big")
+    return identifier & ((1 << ID_LENGTH_BITS) - 1)
+
+
+def generate_mnemonics(
+    group_threshold, groups, master_secret, passphrase=b"", iteration_exponent=0
+):
+    """
+    Splits a master secret into mnemonic shares using Shamir's secret sharing scheme.
+    :param int group_threshold: The number of groups required to reconstruct the master secret.
+    :param groups: A list of (member_threshold, member_count) pairs for each group, where member_count
+        is the number of shares to generate for the group and member_threshold is the number of members required to
+        reconstruct the group secret.
+    :type groups: List of pairs of integers.
+    :param master_secret: The master secret to split.
+    :type master_secret: Array of bytes.
+    :param passphrase: The passphrase used to encrypt the master secret.
+    :type passphrase: Array of bytes.
+    :param int iteration_exponent: The iteration exponent.
+    :return: List of mnemonics.
+    :rtype: List of byte arrays.
+    """
+
+    identifier = _generate_random_identifier()
+
+    if len(master_secret) * 8 < MIN_STRENGTH_BITS:
+        raise ValueError(
+            "The length of the master secret ({} bytes) must be at least {} bytes.".format(
+                len(master_secret), bits_to_bytes(MIN_STRENGTH_BITS)
+            )
+        )
+
+    if len(master_secret) % 2 != 0:
+        raise ValueError(
+            "The length of the master secret in bytes must be an even number."
+        )
+
+    if group_threshold > len(groups):
+        raise ValueError(
+            "The requested group threshold ({}) must not exceed the number of groups ({}).".format(
+                group_threshold, len(groups)
+            )
+        )
+
+    if any(
+        member_threshold == 1 and member_count > 1
+        for member_threshold, member_count in groups
+    ):
+        raise ValueError(
+            "Creating multiple member shares with member threshold 1 is not allowed. Use 1-of-1 member sharing instead."
+        )
+
```

Comment:
You may want to check that the splitting scheme doesn't have "spider form". That is to check that either there is a single group or there is a group with at least two members.

Reply:
I don't know, if the user explicitly asks for it to be this way, then maybe we should let them have it. Could be related to https://github.com/trezor/trezor-firmware/pull/85/files#r280716405

Reply:
The "fork form" (as the opposite of the "spider form") will be probably enforced (or suggested) by the implementation of user interface. Which is probably enough.